src/ucp/core/ucp_listener.c

```diff
@@ -90,63 +104,81 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
-                                                       const void *conn_priv_data,
-                                                       size_t length)
+ucs_status_t
+ucp_listener_ep_create(ucp_listener_h                    listener,
+                       const ucp_wireup_sockaddr_priv_t *client_data,
+                       ucp_listener_accept_t            *accept)
+{
+    ucp_ep_h     ep;
+    ucs_status_t status = ucp_ep_create_accept(listener->wiface.worker,
+                                               client_data, &ep);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    accept->listener = listener;
+    accept->is_ep    = 1;
+    accept->ep       = ep;
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_listener_ep_addr_create(ucp_listener_h listener, void *id,
+                            const ucp_wireup_sockaddr_priv_t *client_data,
+                            size_t length, ucp_listener_accept_t *accept)
+{
+    accept->listener    = listener;
+    accept->is_ep       = 0;
+    accept->addr        = ucs_malloc(ucs_offsetof(ucp_ep_address_t, priv_addr) +
+                                     length, "accept ep addr");
+    if (accept->addr == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    accept->addr->listener    = listener;
+    accept->addr->conn_req_id = id;
+    memcpy(&accept->addr->priv_addr, client_data, length);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_listener_conn_request_callback(void *arg, void *id,
+                                   const void *conn_priv_data, size_t length)
 {
     const ucp_wireup_sockaddr_priv_t *client_data = conn_priv_data;
     ucp_listener_h listener                       = arg;
-    ucp_unpacked_address_t remote_address;
     ucp_listener_accept_t *accept;
     uct_worker_cb_id_t prog_id;
-    ucp_ep_params_t params;
     ucs_status_t status;
-    ucp_ep_h ep;
 
     ucs_trace("listener %p: got connection request", listener);
 
-    params.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
-    params.err_mode   = client_data->err_mode;
-
-    status = ucp_address_unpack(client_data + 1, &remote_address);
-    if (status != UCS_OK) {
-        goto err;
+    /* Defer wireup init and user's callback to be invoked from the main thread */
+    accept = ucs_malloc(sizeof(*accept), "ucp_listener_accept");
+    if (accept == NULL) {
+        ucs_error("failed to allocate listener accept context");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    if (client_data->is_full_addr) {
-        /* create endpoint to the worker address we got in the private data */
-        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
-                                              &remote_address, UCP_EP_CREATE_AM_LANE,
-                                              "listener", &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
+    if (listener->cb) {
+        status = ucp_listener_ep_create(listener, client_data, accept);
```

Comment:
Good question... Since we have 2 types of accept callbacks and the new one does not create EP until user calls `ucp_ep_create`. We don't need to initiate wireup because we don't know on which worker. On other hand, user can set `UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER` flag and pass `NULL`, let's say, to ignore incoming EPs. I think this is not valid usage because it will cause memory leak, user should call `ucp_ep_destroy` for created EPs on server.

Reply:
but for the current flow - before the changes in this pr, the wireup would start even if there is no listener's callback, why should this change? (if neither UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER nor the new flag is set)

Reply:
ep is connected to ep for client-server flow. Even for the current flow, UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER should be mandatory to avoid memory leak related to EP creation on server side. User should handle connection request and destroy the EP after communications.

Reply:
It is not mentioned in the api that this flag is mandatory. Also, if this flag won't be set, in the "worst" case, the ep on the server side would be destoyed as part of the worker's destroy flow.

Reply:
> the ep on the server side would be destoyed as part of the worker's destroy flow

For long living server application this is a memory leak. But anyway I have to return initial behavior for backward compatibility. So, if user does not setup any callback, the server will create internal EP, which can be destroyed only during worker destroy (with current API)

Reply:
Fixed